examples/pytorch/gat/main.py

```diff
@@ -92,58 +52,81 @@ def train_func(config: Dict):
         optimizer, named_parameters=model.named_parameters()
     )
 
-    address = "localhost:9999"
-    s = Server(address, args.data_dir, 0, len(args.partitions))
-    g = DistributedClient([address])
-    max_id = g.node_count(args.node_type) if args.max_id in [-1, None] else args.max_id
-    dataset = ray.data.range(max_id).repartition(max_id // args.batch_size)
-    pipe = dataset.window(blocks_per_window=4).repeat(args.num_epochs)
+    g = config["get_graph"]()
```

Comment:
Config is typically used for passing numerical values, not objects..
I suggest to pass graph directly to the training function.

Reply:
config is the only way to pass values to the training function and they all have to be serializable, it is easier to serialize the function than a GE.

Reply:
why can't we pass [a dataset](https://github.com/ray-project/ray/blob/02ca4c9f2b2311e3fdf47de99303f7473170e477/python/ray/train/torch/torch_trainer.py#L250) for example?

Reply:
We can but we'd have to move a lot to the main function,
```
    g = config["get_graph"]()
    batch_size = 140
    train_dataset = ray.data.read_text(f"{config['data_dir']}/train.nodes")
    train_dataset = train_dataset.repartition(train_dataset.count() // batch_size)
    train_pipe = train_dataset.window(blocks_per_window=4).repeat(config["num_epochs"])

    def transform_batch(idx: list) -> dict:
        return model.q.query_training(g, np.array(idx))

    train_pipe = train_pipe.map_batches(transform_batch)

    test_dataset = ray.data.read_text(f"{config['data_dir']}/test.nodes")
    test_dataset = test_dataset.repartition(1)
    test_dataset = test_dataset.map_batches(transform_batch)
    test_dataset_iter = test_dataset.repeat(config["num_epochs"]).iter_epochs()
```

I can just make the graph serializeable it wont affect anything though?

Reply:
Can we consume a graph client as a ray dataset and then apply pipeline transformations on top of it?

Reply:
The problem is, the graph does not get used until the very last step, so no.

Reply:
why is it a problem?

Reply:
I'm just not sure the idea you are talking about would look like.

Reply:
https://docs.ray.io/en/latest/ray-air/check-ingest.html#how-do-i-pass-in-a-datasetpipeline-to-my-trainer

Reply:
No the "consume graph client as ray dataset" part.

Reply:
Well, you can use [map_batches operation](https://docs.ray.io/en/latest/data/api/dataset.html#ray.data.Dataset.map_batches). In that case query is the transform function and graph client is a constructor argument to the dataset. Will it work?

Reply:
I can maybe do, but it might mess with the serialization of the graph
```
model.graph = config["get_graph"]()
train_pipe = train_pipe.map_batches(model.q.query_training)
```

Reply:
> it might mess with the serialization of the graph

How? graph clients are serializable

Reply:
This is what it will look like overall, good?
```
    model.g = config["get_graph"]()
    batch_size = 140
    train_dataset = ray.data.read_text(f"{config['data_dir']}/train.nodes")
    train_dataset = train_dataset.repartition(train_dataset.count() // batch_size)
    train_pipe = train_dataset.window(blocks_per_window=4).repeat(config["num_epochs"])

    train_pipe = train_pipe.map_batches(model.q.query_training)

    test_dataset = ray.data.read_text(f"{config['data_dir']}/test.nodes")
    test_dataset = test_dataset.repartition(1)
    test_dataset = test_dataset.map_batches(transform_batch)
    test_dataset_iter = test_dataset.repeat(config["num_epochs"]).iter_epochs()
```

Reply:
No, `model.g = config["get_graph"]()` graph is a parameter to a query, not a field for a model.. So it has to be an input too a dataset a model is trained on, not the graph, like I said above: 
`graph client is a constructor argument to the dataset.`

Reply:
Sure, that works:
```python
train_pipe = train_pipe.map_batches(lambda idx: model.q.query_training(g, idx))
```